<img src="../../predictioNN_Logo_JPG(72).jpg" width=200>

---

## Logistic Regression with Breast Cancer Data

### Introduction to Data Science
#### Last Updated: February 24, 2023
---  

### SOURCES
- [Logistic regression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html)

### OBJECTIVES
- Implement logistic regression using `sklearn`
- Use the sigmoid function to compute the predicted probability
- Compute binary classification metrics

### CONCEPTS
- logistic regression

---


## 1. Logistic Regression with `sklearn`

We worked with the sigmoid function for computing probabilities of a binary outcome.  
Logistic regression is a model that does these things:
- Use a linear combination of predictors as input, equal to: $\beta_0 + \beta_1 X_1 + ... + \beta_n X_n$
- Feed the input into the sigmoid function (a.k.a. logistic function)
- Estimate the parameters to minimize error
- Output a probability estimate of the outcome

Now we will work with the Wisconsin Breast Cancer Dataset to predict if a cell is benign ('B') or malignant ('M'). 

The dataset was sourced here:  
https://archive.ics.uci.edu/ml/datasets/breast+cancer+wisconsin+(diagnostic)

In [1]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

In [2]:
datapath = '../datasets/wdbc.csv'

**Read in the Data**

In [3]:
df = pd.read_csv(datapath)
df.head()

id diagnosis     f1     f2      f3      f4       f5       f6      f7  \
0    842302         M  17.99  10.38  122.80  1001.0  0.11840  0.27760  0.3001   
1    842517         M  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.0869   
2  84300903         M  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.1974   
3  84348301         M  11.42  20.38   77.58   386.1  0.14250  0.28390  0.2414   
4  84358402         M  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.1980   

        f8  ...    f21    f22     f23     f24     f25     f26     f27     f28  \
0  0.14710  ...  25.38  17.33  184.60  2019.0  0.1622  0.6656  0.7119  0.2654   
1  0.07017  ...  24.99  23.41  158.80  1956.0  0.1238  0.1866  0.2416  0.1860   
2  0.12790  ...  23.57  25.53  152.50  1709.0  0.1444  0.4245  0.4504  0.2430   
3  0.10520  ...  14.91  26.50   98.87   567.7  0.2098  0.8663  0.6869  0.2575   
4  0.10430  ...  22.54  16.67  152.20  1575.0  0.1374  0.2050  0.4000  0.1625   

      f29      f30  
0  0.4601  0.11890  
1  0.2750  0.08902  
2  0.3613  0.08758  
3  0.6638  0.17300  
4  0.2364  0.07678  

[5 rows x 32 columns]

**About the Fields**

`id` - unique identifier for each subject  
`diagnosis` - target variable indicating if cell is malignant or benign  
`f1-f30` - cell measurement variables

---

**Preprocessing**

The `diagnosis` field is the target variable. It needs to be converted to values of 0 and 1.  
We can make malignant = 1, benign = 0.

In [4]:
df['target'] = df['diagnosis'].apply(lambda x: 1 if x == 'M' else 0)

Let's make sure this is correct

In [5]:
print(df.head())
print(df.tail())

         id diagnosis     f1     f2      f3      f4       f5       f6      f7  \
0    842302         M  17.99  10.38  122.80  1001.0  0.11840  0.27760  0.3001   
1    842517         M  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.0869   
2  84300903         M  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.1974   
3  84348301         M  11.42  20.38   77.58   386.1  0.14250  0.28390  0.2414   
4  84358402         M  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.1980   

        f8  ...    f22     f23     f24     f25     f26     f27     f28  \
0  0.14710  ...  17.33  184.60  2019.0  0.1622  0.6656  0.7119  0.2654   
1  0.07017  ...  23.41  158.80  1956.0  0.1238  0.1866  0.2416  0.1860   
2  0.12790  ...  25.53  152.50  1709.0  0.1444  0.4245  0.4504  0.2430   
3  0.10520  ...  26.50   98.87   567.7  0.2098  0.8663  0.6869  0.2575   
4  0.10430  ...  16.67  152.20  1575.0  0.1374  0.2050  0.4000  0.1625   

      f29      f30  target  
0  0.4601  0.11890       1  
1  0.2750 

**Prepare the data, using f1 and f2 as predictors**

In [6]:
X = df[['f1','f2']].values
y = df['target'].values

**Split into training set, test set**

In [7]:
x_tr, x_te, y_tr, y_te = train_test_split(X, y, train_size = 0.6, random_state=314)

**Print some of the data**

In [8]:
print('x_tr: \n', x_tr[:5,:])
print('')
print('y_tr: \n', y_tr[:5])

x_tr: 
 [[19.21 18.57]
 [19.59 25.  ]
 [10.29 27.61]
 [13.85 19.6 ]
 [12.47 18.6 ]]

y_tr: 
 [1 1 0 0 0]


**Fit the logistic regression model**

In [9]:
model = LogisticRegression().fit(x_tr, y_tr)

**Extract the model coefficients**

In [10]:
model.coef_

array([[1.08461051, 0.22651908]])

**Extract the model intercept**

In [11]:
model.intercept_

array([-20.4275731])

This tells us that the coefficient on f1 and f2 is 1.08461051 and 0.22651908, respectively

**Predict the Cell Types**

In [12]:
# store the predictions for all subjects in a new column
df['label_predicted'] = model.predict(X)

# print the predictions
print(model.predict(X))

[1 1 1 0 1 0 1 0 0 0 1 1 1 1 0 1 1 1 1 0 0 0 0 1 1 1 1 1 1 1 1 0 1 1 1 1 0
 0 1 0 0 0 1 0 0 1 0 0 0 0 0 0 0 1 1 0 1 1 0 0 0 0 1 0 0 1 0 0 0 0 1 0 1 0
 0 1 0 1 1 0 0 0 1 1 0 1 1 1 0 0 1 1 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0
 0 0 0 0 0 0 0 1 1 0 1 1 0 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 0 0 1 0 0 0 0 0 1
 0 0 0 0 0 0 0 0 1 1 0 0 0 1 1 0 1 1 0 1 1 0 0 0 0 0 0 0 0 1 0 0 1 1 1 0 1
 0 1 0 0 0 1 0 0 0 1 0 0 1 1 0 0 1 1 0 0 0 0 1 0 0 1 0 1 1 1 0 0 0 1 1 0 0
 0 1 0 0 0 0 0 0 1 0 0 1 0 0 1 1 1 1 0 0 0 0 1 0 0 0 0 0 1 0 1 1 1 0 1 1 1
 1 1 1 1 1 1 1 0 0 0 0 0 0 1 0 1 0 0 1 0 0 1 0 1 1 0 0 0 0 0 0 0 1 0 0 0 0
 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 1 1 1 0 0
 0 0 1 0 1 0 1 0 0 0 1 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 1 0 1 1 0 1 1
 1 0 1 1 0 1 0 1 0 0 0 0 0 0 0 1 0 0 0 1 0 0 1 1 0 0 0 0 0 0 1 0 0 0 0 0 1
 0 1 0 0 0 0 1 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 1 0 0 0 0 0 0 0 1 0 0
 1 0 1 0 0 1 0 1 0 0 0 1 0 0 0 0 1 1 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0
 0 0 0 0 0 0 1 0 1 0 1 1 

**Filter the dataframe to show subjects where the prediction matches the target. These are correct predictions.**

In [ ]:
df[df['label_predicted'] == df['target']]

**Predict the Probability of Each Cell Type in Training Set**

In [13]:
model.predict_proba(x_tr)[:5,:]

array([[0.00981234, 0.99018766],
       [0.001527  , 0.998473  ],
       [0.95314633, 0.04685367],
       [0.72431022, 0.27568978],
       [0.93638784, 0.06361216]])

For example, for the first subject, the probability of a benign cell is 0.00981234
The probability of a malignant cell is 0.99018766

Since the malignant probability is greater than the default threshold of 0.5, the predicted cell type is 1 (malignant).

**Extracting Positive Probabilities**

It can be useful to extract the probabilities of the positive label for each subject.  
This can be done by extracting the second index across all rows like this:

In [14]:
model.predict_proba(x_tr)[:,1]

array([9.90187665e-01, 9.98473003e-01, 4.68536694e-02, 2.75689785e-01,
       6.36121602e-02, 6.64490382e-01, 6.08318342e-04, 9.98740086e-01,
       3.00316027e-02, 2.22638228e-04, 1.95496262e-01, 9.69121999e-01,
       9.97352442e-01, 7.23361530e-01, 1.82210018e-01, 9.61437814e-01,
       9.15832160e-01, 9.99180405e-01, 5.27272025e-01, 6.82121502e-04,
       4.00821485e-02, 2.70001375e-01, 4.18960044e-02, 9.92910823e-01,
       4.26049965e-02, 9.72717081e-01, 9.95805052e-01, 2.53538062e-01,
       9.99645994e-01, 9.87222697e-02, 9.99720529e-01, 6.97748324e-01,
       4.46193904e-02, 9.72613518e-01, 2.76690213e-02, 1.61160956e-01,
       1.26511261e-01, 3.89033693e-02, 5.84287849e-02, 3.72071168e-01,
       6.79340036e-01, 5.88185506e-01, 1.03313627e-02, 2.46226871e-02,
       9.85902604e-01, 9.51332119e-01, 9.99635763e-01, 7.16450651e-03,
       7.69060702e-02, 1.71056600e-02, 2.60862252e-02, 9.77729619e-01,
       1.04673106e-02, 7.39463454e-01, 8.94537086e-02, 8.10493161e-01,
      

Suppose we want to change the threshold, predicting cell type 1 if the probability of the positive label is greater than 0.85. This now requires a higher confidence compared to using 0.5 as the threshold.

In [15]:
model.predict_proba(x_tr)[:,1] > 0.85

array([ True,  True, False, False, False, False, False,  True, False,
       False, False,  True,  True, False, False,  True,  True,  True,
       False, False, False, False, False,  True, False,  True,  True,
       False,  True, False,  True, False, False,  True, False, False,
       False, False, False, False, False, False, False, False,  True,
        True,  True, False, False, False, False,  True, False, False,
       False, False,  True, False, False, False,  True,  True, False,
       False,  True, False, False,  True, False, False, False,  True,
        True,  True, False, False, False, False, False, False, False,
        True, False,  True, False,  True, False,  True, False,  True,
       False, False, False, False, False, False,  True, False, False,
       False, False, False, False, False,  True,  True, False, False,
       False, False,  True, False, False,  True, False, False, False,
       False, False, False,  True, False,  True, False,  True, False,
       False, False,

**Measuring Performance**

In [33]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import precision_recall_curve

In [34]:
confusion_matrix(y_te, model.predict(x_te))

array([[141,  10],
       [ 19,  58]], dtype=int64)

Compute batch of metrics

In [35]:
print(classification_report(y_te, model.predict(x_te)))

              precision    recall  f1-score   support

           0       0.88      0.93      0.91       151
           1       0.85      0.75      0.80        77

    accuracy                           0.87       228
   macro avg       0.87      0.84      0.85       228
weighted avg       0.87      0.87      0.87       228



Compute metrics over range of thresholds

In [64]:
# probabilities of positive class from test set
pred_pos = model.predict_proba(x_te)[:,1]

# calculate metrics over range of thresholds
pre, re, th = precision_recall_curve(y_te,  pred_pos)

# print metrics for a sample of thresholds
print('precision:', pre[100:105])
print('')
print('recall:   ', re[100:105])
print('')
print('threshold:', th[100:105])

precision: [0.87692308 0.890625   0.9047619  0.90322581 0.90163934]

recall:    [0.74025974 0.74025974 0.74025974 0.72727273 0.71428571]

threshold: [0.52326871 0.53113831 0.53937435 0.55367587 0.563716  ]


Notice that as threshold increases:
- precision increases (fewer false positives)
- recall decreases (more false negatives)

This is the tradeoff in classification problems

---

**THINK ABOUT AND DISCUSS**

1) If you raise the threshold for predicting a positive label, what generally happens to the recall? What generally happens to the precision?

answer: Since we require a higher confidence to predict the positive label, the precision will likely increase.  However, the recall will likely decrease, producing more false negatives. This is why it is important to measure precision and recall together.

---

**TRY FOR YOURSELF**

You will evaluate the model performance.

Hint: you can count the number of rows in a dataframe by using the `len()` function like this: `len(df)`

2) Compute the accuracy of the model, where accuracy = #correct / #total

In [ ]:
# answer
len(df[df['label_predicted'] == df['target']]) / len(df)

3) Count the number of true positives (where the model predicted 1 and the target is 1)

In [ ]:
# answer
len(df[ (df['label_predicted'] == 1) & (df['target'] == 1) ])

4) Compute the recall of the model, where precision = #true_positive / #actual_positive

In [ ]:
# answer
len(df[ (df['label_predicted'] == 1) & (df['target'] == 1) ]) / len( df[df['target'] == 1])

---

5) As we saw earlier, the first subject in the training set has a probability of malignancy = 0.99018766  
Compute this by using the intercept, coefficients, f1, f2 values, and the definition of the sigmoid:

$sigmoid = 1 / ( 1 + np.exp(-(b0 + b1 * x1 + b2 * x2) ))$

Note this version uses two predictors.

In [16]:
#answer
b0 = model.intercept_
b1 = model.coef_[0][0]
b2 = model.coef_[0][1]
x1 = x_tr[0][0]
x2 = x_tr[0][1]

1 / ( 1 + np.exp(-(b0 + b1 * x1 + b2 * x2) ))

array([0.99018766])

---